File: ch04/top-down-parser/parser.c
/*
  parser.c - a recursive descent parser for simple arithmetic expressions

  identifier: a sequence of letters and digits, starting with a letter
  integer literal (number): a sequence of digits
  operator: +, -, *, /
  parentheses: (, )

  Grammar (EBNF):
    expr = term, { ("+" | "-"), term } ;
    term = factor, { ("*" | "/"), factor } ;
    factor = identifier | integer | "(" expr ")" ;
    identifier = letter, { letter | digit } ;
    integer = digit, { digit } ;
    letter = "A" | "B" | "C" | ... | "Z" | "a" | "b" | "c" | ... | "z" ;
    digit = "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9" ;

 */

#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>

/* Global declarations */
/* Variables */
int charClass;
char lexeme[100];
char nextChar;
int lexLen;
int token;
int nextToken;
FILE *in_fp;

/* Function declarations */
void addChar();
void getChar();
void getNonBlank();
int lex();
int lookup(char ch);
void error(char *message);

/* Parser function declarations */
void expr();
void term();
void factor();

/* Character classes */
#define LETTER 0
#define DIGIT 1
#define UNKNOWN 99

/* Token codes */
#define INT_LIT 10
#define IDENT 11
#define ASSIGN_OP 20
#define ADD_OP 21
#define SUB_OP 22
#define MULT_OP 23
#define DIV_OP 24
#define LEFT_PAREN 25
#define RIGHT_PAREN 26

/******************************************************/
/* main driver */
int main() {
    /* Open the input data file and process its contents */
    if ((in_fp = fopen("front.in", "r")) == NULL) {
        printf("ERROR - cannot open front.in \n");
        return 1;
    }

    /* Get the first character of input */
    getChar();

    /* Get the first token */
    lex();

    /* Parse the expression */
    expr();

    /* Check for end of file */
    if (nextToken != EOF)
        error("Unexpected symbols after end of expression");

    printf("\nParsing completed successfully.\n");

    fclose(in_fp);
    return 0;
}

/*****************************************************/
/* lookup - a function to lookup operators and parentheses
   and return the token */
int lookup(char ch) {
    switch (ch) {
        case '(':
            addChar();
            nextToken = LEFT_PAREN;
            break;
        case ')':
            addChar();
            nextToken = RIGHT_PAREN;
            break;
        case '+':
            addChar();
            nextToken = ADD_OP;
            break;
        case '-':
            addChar();
            nextToken = SUB_OP;
            break;
        case '*':
            addChar();
            nextToken = MULT_OP;
            break;
        case '/':
            addChar();
            nextToken = DIV_OP;
            break;
        default:
            addChar();
            nextToken = EOF;
            break;
    }
    return nextToken;
}

/*****************************************************/
/* addChar - a function to add nextChar to lexeme */
void addChar() {
    if (lexLen <= 98) {
        lexeme[lexLen++] = nextChar;
        lexeme[lexLen] = 0;
    }
    else
        printf("Error - lexeme is too long \n");
}

/*****************************************************/
/* getChar - a function to get the next character of
   input and determine its character class */
void getChar() {
    if ((nextChar = getc(in_fp)) != EOF) {
        if (isalpha(nextChar))
            charClass = LETTER;
        else if (isdigit(nextChar))
            charClass = DIGIT;
        else charClass = UNKNOWN;
    }
    else
        charClass = EOF;
}

/*****************************************************/
/* getNonBlank - a function to call getChar until it
   returns a non-whitespace character */
void getNonBlank() {
    while (isspace(nextChar))
        getChar();
}

/*****************************************************/
/* lex - a simple lexical analyzer for arithmetic
   expressions */
int lex() {
    lexLen = 0;
    getNonBlank();

    switch (charClass) {
        /* Parse identifiers */
        case LETTER:
            addChar();
            getChar();
            while (charClass == LETTER || charClass == DIGIT) {
                addChar();
                getChar();
            }
            nextToken = IDENT;
            break;

        /* Parse integer literals */
        case DIGIT:
            addChar();
            getChar();
            while (charClass == DIGIT) {
                addChar();
                getChar();
            }
            nextToken = INT_LIT;
            break;

        /* Parentheses and operators */
        case UNKNOWN:
            lookup(nextChar);
            getChar();
            break;

        /* EOF */
        case EOF:
            nextToken = EOF;
            lexeme[0] = 'E';
            lexeme[1] = 'O';
            lexeme[2] = 'F';
            lexeme[3] = 0;
            break;
    }

    printf("Next token is: %d, Next lexeme is %s\n",
        nextToken, lexeme);
    return nextToken;
}

/*****************************************************/
/* expr
   Parses strings in the language generated by the rule:
   <expr> -> <term> {(+ | -) <term>}
   */
void expr() {
    printf("Enter <expr>\n");

    /* Parse the first term */
    term();

    /* As long as the next token is + or -, get
       the next token and parse the next term */
    while (nextToken == ADD_OP || nextToken == SUB_OP) {
        lex();
        term();
    }

    printf("Exit <expr>\n");
}

/*****************************************************/
/* term
   Parses strings in the language generated by the rule:
   <term> -> <factor> {(* | /) <factor>}
   */
void term() {
    printf("Enter <term>\n");

    /* Parse the first factor */
    factor();

    /* As long as the next token is * or /, get the
       next token and parse the next factor */
    while (nextToken == MULT_OP || nextToken == DIV_OP) {
        lex();
        factor();
    }

    printf("Exit <term>\n");
}

/*****************************************************/
/* factor
   Parses strings in the language generated by the rule:
   <factor> -> id | int_constant | ( <expr> )
   */
void factor() {
    printf("Enter <factor>\n");

    /* Determine which RHS */
    if (nextToken == IDENT || nextToken == INT_LIT) {
        /* Get the next token */
        lex();
    }
    /* If the RHS is ( <expr> ), call lex to pass over the
       left parenthesis, call expr, and check for the right
       parenthesis */
    else if (nextToken == LEFT_PAREN) {
        lex();
        expr();
        if (nextToken == RIGHT_PAREN)
            lex();
        else
            error("Right parenthesis expected");
    }
    else
        error("Expected identifier, integer, or left parenthesis");

    printf("Exit <factor>\n");
}

/*****************************************************/
/* error - prints error message and exits */
void error(char *message) {
    printf("Error: %s\n", message);
    exit(1);
}

File: ch04/top-down-parser/parser.py
import sys
from enum import Enum, auto

# =============================================================================
# 1. TOKENS
# =============================================================================

class TokenType(Enum):
    INT_LIT = auto()       # Integer literals (e.g., 10, 42)
    IDENT = auto()         # Identifiers (e.g., sum, total)
    ASSIGN_OP = auto()     # =
    ADD_OP = auto()        # +
    SUB_OP = auto()        # -
    MULT_OP = auto()       # *
    DIV_OP = auto()        # /
    LEFT_PAREN = auto()    # (
    RIGHT_PAREN = auto()   # )
    EOF = auto()           # End of File

class Token:
    def __init__(self, type_: TokenType, value: str):
        self.type = type_
        self.value = value

    def __repr__(self):
        return f"Token({self.type.name}, {repr(self.value)})"

# =============================================================================
# 2. ABSTRACT SYNTAX TREE (AST) NODES
# =============================================================================

class AST:
    """Base class for all AST nodes."""
    pass

class BinOp(AST):
    """Represents a binary operation (e.g., left + right)."""
    def __init__(self, left, op, right):
        self.left = left
        self.token = self.op = op
        self.right = right

    def __repr__(self):
        # Format: (left op right)
        return f"({self.left} {self.op.value} {self.right})"

class Num(AST):
    """Represents an integer number."""
    def __init__(self, token):
        self.token = token
        self.value = token.value

    def __repr__(self):
        return f"{self.value}"

class Var(AST):
    """Represents a variable/identifier."""
    def __init__(self, token):
        self.token = token
        self.value = token.value

    def __repr__(self):
        return f"Var({self.value})"

# =============================================================================
# 3. LEXER
# =============================================================================

class Lexer:
    def __init__(self, text):
        self.text = text
        self.pos = 0
        self.current_char = self.text[self.pos] if self.text else None

    def error(self, msg="Invalid character"):
        raise Exception(f"Lexer Error: {msg}")

    def advance(self):
        """Advance the 'pos' pointer and set the 'current_char' variable."""
        self.pos += 1
        if self.pos > len(self.text) - 1:
            self.current_char = None  # Indicates End of Input
        else:
            self.current_char = self.text[self.pos]

    def skip_whitespace(self):
        while self.current_char is not None and self.current_char.isspace():
            self.advance()

    def integer(self):
        """Return a (multidigit) integer consumed from the input."""
        result = ''
        while self.current_char is not None and self.current_char.isdigit():
            result += self.current_char
            self.advance()
        return Token(TokenType.INT_LIT, result)

    def identifier(self):
        """Handle identifiers (variables)."""
        result = ''
        while self.current_char is not None and self.current_char.isalnum():
            result += self.current_char
            self.advance()
        return Token(TokenType.IDENT, result)

    def get_next_token(self):
        """Lexical analyzer (also known as scanner or tokenizer)"""
        while self.current_char is not None:

            if self.current_char.isspace():
                self.skip_whitespace()
                continue

            if self.current_char.isalpha():
                return self.identifier()

            if self.current_char.isdigit():
                return self.integer()

            if self.current_char == '+':
                self.advance()
                return Token(TokenType.ADD_OP, '+')

            if self.current_char == '-':
                self.advance()
                return Token(TokenType.SUB_OP, '-')

            if self.current_char == '*':
                self.advance()
                return Token(TokenType.MULT_OP, '*')

            if self.current_char == '/':
                self.advance()
                return Token(TokenType.DIV_OP, '/')

            if self.current_char == '(':
                self.advance()
                return Token(TokenType.LEFT_PAREN, '(')

            if self.current_char == ')':
                self.advance()
                return Token(TokenType.RIGHT_PAREN, ')')

            self.error(f"Unknown character: {self.current_char}")

        return Token(TokenType.EOF, None)

# =============================================================================
# 4. PARSER (UPDATED TO BUILD AST)
# =============================================================================

class Parser:
    def __init__(self, lexer):
        self.lexer = lexer
        self.current_token = self.lexer.get_next_token()

    def error(self, msg="Invalid syntax"):
        raise Exception(f"Parser Error: {msg}")

    def eat(self, token_type):
        """
        Compare the current token type with the passed token type.
        If they match, 'eat' the current token and advance to the next.
        """
        if self.current_token.type == token_type:
            self.current_token = self.lexer.get_next_token()
        else:
            self.error(f"Expected {token_type.name}, got {self.current_token.type.name}")

    def factor(self):
        """
        factor : IDENT | INT_LIT | LPAREN expr RPAREN
        """
        token = self.current_token

        if token.type == TokenType.INT_LIT:
            self.eat(TokenType.INT_LIT)
            return Num(token)  # Return a Leaf Node (Number)

        elif token.type == TokenType.IDENT:
            self.eat(TokenType.IDENT)
            return Var(token)  # Return a Leaf Node (Variable)

        elif token.type == TokenType.LEFT_PAREN:
            self.eat(TokenType.LEFT_PAREN)
            node = self.expr() # Recurse
            self.eat(TokenType.RIGHT_PAREN)
            return node        # Return the sub-tree result

        self.error("Expected identifier, integer, or left parenthesis")

    def term(self):
        """
        term : factor ((MULT | DIV) factor)*
        """
        # 1. Get the left-hand side
        node = self.factor()

        # 2. Iterate while we see * or /
        while self.current_token.type in (TokenType.MULT_OP, TokenType.DIV_OP):
            token = self.current_token
            if token.type == TokenType.MULT_OP:
                self.eat(TokenType.MULT_OP)
            elif token.type == TokenType.DIV_OP:
                self.eat(TokenType.DIV_OP)

            # 3. Create a BinOp node:
            #    Left child = the 'old' node
            #    Right child = the result of the next factor()
            node = BinOp(left=node, op=token, right=self.factor())

        return node

    def expr(self):
        """
        expr : term ((PLUS | MINUS) term)*
        """
        # 1. Get the left-hand side
        node = self.term()

        # 2. Iterate while we see + or -
        while self.current_token.type in (TokenType.ADD_OP, TokenType.SUB_OP):
            token = self.current_token
            if token.type == TokenType.ADD_OP:
                self.eat(TokenType.ADD_OP)
            elif token.type == TokenType.SUB_OP:
                self.eat(TokenType.SUB_OP)

            # 3. Create a BinOp node
            node = BinOp(left=node, op=token, right=self.term())

        return node

    def parse(self):
        """Main entry point."""
        if self.current_token.type == TokenType.EOF:
            return None

        result = self.expr()

        if self.current_token.type != TokenType.EOF:
            self.error("Unexpected symbols after end of expression")

        return result

# =============================================================================
# 5. MAIN DRIVER
# =============================================================================

def main():
    # Read expression from standard input
    input_text = input().strip()

    print(f"Input Expression: {input_text}")
    print("-" * 30)

    lexer = Lexer(input_text)
    parser = Parser(lexer)

    try:
        syntax_tree = parser.parse()
        print("Generated Parse Tree (AST):")
        print(syntax_tree)
    except Exception as e:
        print(f"Error: {e}")

if __name__ == "__main__":
    main()
